nmigen/build/generic_platform.py

```diff
@@ -0,0 +1,374 @@
+                elif sub.io[0].dir == "o":
+                    shape = len(sub.io[0])
+                    direction = DIR_FANOUT
+                else:
+                    shape = _triple_layout(len(sub.io[0]))
+                    direction = DIR_NONE
+                fields.append((sub.name, shape, direction))
+        super().__init__(fields)
+
+
+class ConstraintError(Exception):
+    pass
+
+
+class _ConstraintManager:
+    def __init__(self, resources, connectors):
+        self.available    = defaultdict(dict)
+        self.matched      = OrderedDict()
+        self.connectors   = dict()
+        self.commands     = []
+        self.clocks       = OrderedDict()
+        self.subfragments = []
+        self.ports        = []
+        self.constraints  = OrderedDict()
+
+        self.add_resources(resources)
+        self.add_connectors(connectors)
+
+    def add_resources(self, resources):
+        for r in resources:
+            if not isinstance(r, Resource):
+                raise TypeError("Object {!r} is not a Resource".format(r))
+            if r.number in self.available[r.name]:
+                raise NameError("Resource {!r} has a (name, number) pair that is already taken"
+                                .format(r))
+            self.available[r.name][r.number] = r
+
+    def add_connectors(self, connectors):
+        for c in connectors:
+            if not isinstance(c, Connector):
+                raise TypeError("Object {!r} is not a Connector".format(c))
+            if c.name in self.connectors:
+                raise NameError("Connector {!r} has a name that is already taken".format(c))
+            self.connectors[c.name] = c
+
+    def add_command(self, command):
+        if not isinstance(command, str):
+            raise TypeError("Command must be a string, not {!r}".format(command))
+        self.commands.append(command)
+
+    def add_period_constraint(self, resource, period):
+        if not isinstance(resource, Resource):
+            raise TypeError("Object {!r} is not a Resource".format(resource))
+        if (resource.name, resource.number) in self.clocks:
+            raise ConstraintError("Resource {!r} is already constrained to a period"
+                                  .format(resource))
+        self.clocks[resource.name, resource.number] = period
+
+    def lookup(self, name, number=None):
+        if number is None and self.available[name]:
+            number = sorted(self.available[name])[0]
+        elif number not in self.available[name]:
+            raise NameError("No available Resource with name '{}' and number '{}'"
+                            .format(name, number))
+        return self.available[name][number]
+
+    def request(self, name, number=None, dir=None, xdr=1):
+        resource = self.lookup(name, number=number)
+        assert (resource.name, resource.number) not in self.matched
+
+        if dir is not None:
+            resource.update_dir(dir)
```

Comment:
How are resources supposed to be added? If this is done the same way as in oMigen, i.e. you have a global array of `_io` somewhere, then this is completely wrong, because trying to build two different designs in one Python invocation will fail.

In general, nMigen avoids mutating anything, which is slightly slower but eliminates hard-to-find bugs like this.